File: Beautiful-Triplets.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

/*
 * Complete the 'beautifulTriplets' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER d
 *  2. INTEGER_ARRAY arr
 */

int beautifulTriplets(int d, vector<int> arr)
{
    auto check = [&arr](int n)
    {
        return std::binary_search(begin(arr), end(arr), n);
    };

    auto result = std::count_if( begin(arr), end(arr),
    [&](int n)
    {
        return check(d+n) && check(2*d + n);
    });
    
    return result;
}

File: Beautiful-Triplets.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

class Result {

    /*
     * Complete the 'beautifulTriplets' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts following parameters:
     *  1. INTEGER d
     *  2. INTEGER_ARRAY arr
     */

    public static int beautifulTriplets(int d, List<Integer> arr) {
        Predicate<Integer> check = n -> arr.contains(n);

        int result = (int) arr.stream()
                            .filter(n -> check.test(d + n) && check.test(2 * d + n))
                            .count();

        return result;
    }

    public static int beautifulTriplets2(int d, List<Integer> arr) {
        Predicate<Integer> check = n -> Collections.binarySearch(arr, n) >= 0;
        //Function<Integer, Boolean> check2 = n -> Collections.binarySearch(arr, n) >= 0;

        int result = 0;
        for (int n : arr) {
            if (check.test(n + d) && check.test(n + 2 * d)) {
                result++;
            }
        }

        return result;        
    }
}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] firstMultipleInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

        int n = Integer.parseInt(firstMultipleInput[0]);

        int d = Integer.parseInt(firstMultipleInput[1]);

        List<Integer> arr = Stream.of(bufferedReader.readLine().replaceAll("\\s+$", "").split(" "))
            .map(Integer::parseInt)
            .collect(toList());

        int result = Result.beautifulTriplets(d, arr);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}
